ocaml/xapi-idl/cluster/cluster_interface.ml

```diff
@@ -15,12 +15,51 @@ type debug_info = string [@@deriving rpcty]
 (** Name of the cluster *)
 type cluster_name = string [@@deriving rpcty]
 
-(** An IPv4 address (a.b.c.d) *)
-type address = IPv4 of string [@@deriving rpcty]
+type ip = IPv4 of string | IPv6 of string [@@deriving rpcty]
 
-let printaddr () = function IPv4 s -> Printf.sprintf "IPv4(%s)" s
+let string_of_ip = function IPv4 a | IPv6 a -> a
 
-let str_of_address address = match address with IPv4 a -> a
+(** this address includes the hostname and hostuuid along with an ip address
+  this is done to maintain backwards compatability, and should be combined with
+  the other variant in the future.
+  *)
+type extended_addr = {ip: ip; hostuuid: string; hostname: string}
+[@@deriving rpcty]
+
+type address = IPv4 of string | Extended of extended_addr [@@deriving rpcty]
+
+let ipstr_of_address = function
+  | IPv4 a ->
+      a
+  | Extended {ip; _} -> (
+    match ip with IPv4 a | IPv6 a -> a
+  )
+
+let ipaddr_of_address = function
+  | IPv4 _ as ip ->
+      ip
+  | Extended _ as extended_addr ->
+      IPv4 (ipstr_of_address extended_addr)
+
+(** The complete address potentially including uuid and hostname *)
+let fullstr_of_address = function
+  | IPv4 a ->
+      a
+  | Extended {ip; hostuuid; hostname} ->
+      Printf.sprintf "(%s, %s, %s)" (string_of_ip ip) hostuuid hostname
+
+let printaddr () = function
```

Comment:
What is the `()` doing here when we an additional argument that we match on?

Reply:
I believe this is to make `%a` to work, but I am not sure why a `()` is used here instead of a `Format.formatter`